[inductor] enable mkldnn op weight pre-packing on aarch64

snadampal opened the pull request:
This PR enables the fx passes and mkldnn optimizations for aarch64 It improved the bert inference performance up to 5.8x on AWS c7g instance when compared torch.compile() vs no compile path. This is enabled when pytorch is built with USE_MKLDNN_ACL option for aarch64.




cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

snadampal commented:
Hi lc0, @lyoka , @nimin98 , minjungkim85, @yns88, appreciate if any of you can review this PR. thank you!

snadampal commented on torch/_inductor/fx_passes/mkldnn_fusion.py:
```diff
@@ -1070,7 +1073,13 @@ def forward(self, x):
     @functools.lru_cache(None)
     def _mkldnn_fusion_init():
-        if torch.backends.mkldnn.enabled and torch.backends.mkldnn.is_available():
+        # TODO: aarch64: enable op fusion for acl once it supports fused operators. Disabling it for now.
```
Hi @jgong5 , one question on operator fusion, my plan was just to enable weights pre-packing on aarch64 so I didn't enable fusion. But it looks like the packing has to go along with the linear + pointwise fusion, because we are saying the packed linear op is linear pointwise, right? without fusion I was expecting the linear op to be rendered separately , but doesn't looks like it's happening .

jgong5 replied:
`mkldnn_linear_pointwise` supports both fusion and non-fusion cases. It accepts the fusion attribution (`attr`) as "none" meaning no fusion.
```c++
static Tensor mkldnn_linear_pointwise(
    const Tensor& input_t,
    const Tensor& weight_t,
    const c10::optional<Tensor>& bias_opt,
    c10::string_view attr, // primitive attribute specifying fusion
    torch::List<c10::optional<at::Scalar>> scalars,
    c10::optional<c10::string_view> algorithm)
```

snadampal replied:
thank you @jgong5 . my main doubt was:
without compile, I see `matmul` and` eltwise_gelu_erf`ops are dispatched separately and with the linear rewrite pass I see packed linear nodes added. This is good so far. But when the compiled graph is dispatched, in fusion case, I see `inner product+gelu_erf` (as postop) but, in non-fusion case, I only see the 'inner product',  but no separate ` eltwise_gelu_erf' for oneDNN. So, I'm trying to understand in non-fusion case, how's gelu_erf handled? is it done inside aten kernels?

jgong5 replied:
The weight repacking passes and the post-op fusion passes are separate passes where weight prepacking passes are executed first (in the freezing stage) followed by the post-op fusion passes. The weight prepacking passes would replace the matmul/linear ops with mkldnn.linear_pointwise. Then the post-op fusion passes would match the patterns with mkldnn.linear_pointwise. If you skip the post-op fusion passes, the "gelu" post-op will be left separately from the linear op and you should notice a separate gelu op in the graph. Did you see that?

snadampal replied:
as I suspected, looks like `gelu` is indeed dispatched via vectorized code gen for non-fusion scenario, looks like it is decomposed to `erf` in the compiled pass, that's why I didn't find references to `gelu` but  found references to `erf` in the compiled graph and also in perf profile. This answers my question about how the gelu was handled when it is not fused in mkldnn ops.

compiled graph (after post grad pass, without mkldnn op fusion):
~~~
 %_linear_pointwise_default_140 : [num_users=2] = call_function[target=torch.ops.mkldnn._linear_pointwise.default](args = (%add_7, %_frozen_param278, %_frozen_param16, none, [], ), kwargs = {})
    %mul_5 : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%_linear_pointwise_default_140, 0.5), kwargs = {})
    %mul_6 : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%_linear_pointwise_default_140, 0.7071067811865476), kwargs = {})
    %erf : [num_users=1] = call_function[target=torch.ops.aten.erf.default](args = (%mul_6,), kwargs = {})
    %add_8 : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%erf, 1), kwargs = {})
    %mul_7 : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%mul_5, %add_8), kwargs = {})
~~~

perf profile:
~~~
     0.82%     0.40%  python3  libm.so.6            [.] erff32
     0.12%     0.02%  python3  [kernel.kallsyms]    [k] perf_iterate_sb
     0.10%     0.00%  python3  [kernel.kallsyms]    [k] perf_event_exit_task
     0.10%     0.00%  python3  [kernel.kallsyms]    [k] perf_event_task
     0.10%     0.00%  python3  [kernel.kallsyms]    [k] perf_event_task_output
     0.07%     0.00%  python3  [kernel.kallsyms]    [k] __perf_event_header__init_id
     0.05%     0.00%  python3  libtorch_python.so   [.] torch::autograd::THPVariable_erf
     0.05%     0.00%  python3  libtorch_cpu.so      [.] at::_ops::erf::call
~~~

snadampal commented:
btw, I had already tested bert sentiment analysis with torch.compile() and the results look correct on aarch64. my above questions were just to better understand the fx_pass and mkldnn rewrite behavior for non-fusion cases, i will dig into the code.

and this PR is ready for merge once approved by the module owners.

snadampal commented:
I have updated the PR to allow the dynamic shapes on aarch64 even for fp32 inputs.

snadampal commented:
Hi lc0, @lyoka , @nimin98 , minjungkim85, @yns88, appreciate if any of you can review this PR. Given the performance gains from torch.compile() it will be really great if it can be merged to PyTorch 2.2. Thank you!

malfet reviewed (approved):
LGTM, but I wonder if `is_mkldnn_acl_supported()` is in any way fundamentally different than `_is_mkldnn_bf16_supported()`? If not, why not merge two functions together?

malfet commented on aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp:
```diff
@@ -27,6 +27,14 @@ static bool is_mkldnn_fp16_supported() {
 }
 
+constexpr bool is_mkldnn_acl_supported() {
+#if defined(__aarch64__) && AT_MKLDNN_ACL_ENABLED()
+  return true;
+#else
+  return false;
+#endif
```
Sorry, I'm wrong, `defined` macro is only valid as part of `#if` , but why `AT_MKLDNN_ACL_ENABLED` is not sufficient?

snadampal replied:
I think  `AT_MKLDNN_ACL_ENABLED` is sufficient, I have updated the PR, will check CI runs to see if there is any break on other platforms.

snadampal committed:
enable mkldnn op weight pre-packing on aarch64

This PR enables the fx passes and mkldnn optimizations for aarch64.
It improved the bert inference performance up to 5.8x on AWS c7g instance
when compared torch.compile() vs no compile path. This is enabled
when pytorch is built with USE_MKLDNN_ACL option on aarch64.
